packages/wallet/src/services/WalletAssetProvider.ts

```diff
@@ -0,0 +1,214 @@
+  for (const assetId of assetIds) {
+    try {
+      const fetchedAsset = await provider.getAsset({
+        assetId,
+        extraData: { nftMetadata: true, tokenMetadata: true }
+      });
+      assetsFromProvider.push(fetchedAsset);
+    } catch (error) {
+      logger.error(error);
+    }
+  }
+
+  return assetsFromProvider;
+};
+
+const createFallbackAsset = (assetId: Cardano.AssetId): Asset.AssetInfo => {
+  const name = Cardano.AssetId.getAssetName(assetId);
+  const policyId = Cardano.AssetId.getPolicyId(assetId);
+  return {
+    assetId,
+    fingerprint: Cardano.AssetFingerprint.fromParts(policyId, name),
+    name,
+    policyId,
+    quantity: 0n,
+    supply: 0n
+  };
+};
+
+const mergeAssets = (
+  assetIds: Cardano.AssetId[],
+  cachedAssetsInfo: Map<Cardano.AssetId, Asset.AssetInfo>,
+  assetsFromProvider: Asset.AssetInfo[],
+  mintedAssets: Asset.AssetInfo[] | null
+): Asset.AssetInfo[] =>
+  assetIds.map((assetId) => {
+    const asset = cachedAssetsInfo.get(assetId) || assetsFromProvider.find((a) => a.assetId === assetId);
+    const mintedAsset = mintedAssets?.find((info) => info.assetId === assetId);
+
+    if (!asset && !mintedAsset) {
+      return createFallbackAsset(assetId);
+    }
+
+    if (!asset && mintedAsset) {
+      return mintedAsset;
+    }
+
+    if (asset && mintedAsset) {
+      asset.supply += mintedAsset.supply;
+
+      if (mintedAsset.nftMetadata) {
+        asset.nftMetadata = mintedAsset.nftMetadata;
+      }
+    }
+
+    return asset!;
+  });
+
+export const createWalletAssetProvider = ({
+  assetProvider,
+  assetInfo$,
+  tx,
+  logger
+}: AssetProviderContext): AssetProvider => ({
+  async getAsset({ assetId }: GetAssetArgs): Promise<Asset.AssetInfo> {
+    const mintedAssets = await getMintedAssetInfosFromTx(tx, logger);
+    const cachedAssetsInfo = await firstValueFrom(assetInfo$);
+
+    let asset = cachedAssetsInfo.get(assetId);
+
+    if (!asset) {
+      try {
+        asset = await assetProvider.getAsset({ assetId, extraData: { nftMetadata: true, tokenMetadata: true } });
+      } catch (error) {
+        logger.error(error);
+      }
+    }
+
+    const mintedAsset = mintedAssets?.find((info) => info.assetId === assetId);
+
+    // Let's create dummy AssetInfo for the unresolved asset. This is probably better than throwing as the UI can still present it as regular token.
+    if (!asset && !mintedAsset) {
+      const name = Cardano.AssetId.getAssetName(assetId);
+      const policyId = Cardano.AssetId.getPolicyId(assetId);
+      return {
+        assetId,
+        fingerprint: Cardano.AssetFingerprint.fromParts(policyId, name),
+        name,
+        policyId,
+        quantity: 0n,
+        supply: 0n
+      };
+    }
+
+    if (!asset) return mintedAsset!;
+
+    if (mintedAsset) {
+      asset.supply += mintedAsset.supply;
+
+      // We give preference to the metadata in the transaction if preset as this would be the most up to date.
```

Comment:
Actually for cip-68 metadata, tokens don't have to be minted to update metadata. We might have a logic error here - we should probably always search for reference tokens in outputs.
```suggestion
    if (mintedAsset) {
      asset.supply += mintedAsset.supply;

      // We give preference to the metadata in the transaction if preset as this would be the most up to date.
```

Reply:
So we need to do the search for all tokens, see if they are CIP-68 and then check again vs the outputs. I will add a final step to do this

Reply:
done here https://github.com/input-output-hk/cardano-js-sdk/pull/1075/commits/eb8eb75ad82bcefdee30286759d3114cfd6f5580